kubernetes: start/stop pods in parallel

a-robinson opened the pull request:
This makes starting a cluster faster, with no real downside for us since
Cockroach has been built to handle multiple nodes joining at a time.

This also takes nodes down in parallel, which isn't something someone
should do if they want their cluster to remain available, but it's
already true that taking multiple nodes down without decommissioning
them will take a cluster offline. This doesn't make that any worse.

Release note: None

------------------

Like #23475, this feature was added in kubernetes v1.8, so we shouldn't switch over to it immediately. We'll also have to update our kubernetes docs if/when this goes in to reflect that all the pods come up (and ask for their certs) in parallel.

a-robinson committed:
kubernetes: start/stop pods in parallel

This makes starting a cluster faster, with no real downside for us since
Cockroach has been built to handle multiple nodes joining at a time.

This also takes nodes down in parallel, which isn't something someone
should do if they want their cluster to remain available, but it's
already true that taking multiple nodes down without decommissioning
them will take a cluster offline. This doesn't make that any worse.

This prevents running on Kubernetes versions 1.7 or older, so I've
updated the docs accordingly.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/23477)
<!-- Reviewable:end -->

smarterclayton commented:
If pod management policy needs a new mode (parallel up, ordered down) we can add that.  I had been waiting for someone to have a really concrete example of letting Kube manage tear down instead of wanting to manage it themselves.

a-robinson commented:
Thanks for chiming in, @smarterclayton! I don't think that ordered tear down would do much for us today. It would only be useful if there was also some sort of "scale-down" hook that would allow us to run a command when the size of statefulset is being decreased. We could then use that hook to tell the cockroach node to drain all of its data onto other cockroach nodes before it shut down, which would allow for safe scaling down of multiple nodes if they were done one-by-one.

As things are today, we don't want to drain all of a node's data every time it gets stopped, since it'd be incredibly wasteful if the node is coming right back up in a matter of seconds or minutes (e.g. due to a rolling restart). That means that as of today when the statefulset is actually scaled down, we don't create new copies of the data that was on that node until a configurable threshold of time passes, at which point we consider the node dead. If multiple nodes are brought down in less than that amount of time (5 minutes by default) without the admin having manually told cockroach to drain the nodes, then unavailability is likely.

Please correct me if I'm wrong about there not being an existing hook for statefulset scale-down, though. That would be very useful for us.

smarterclayton commented:
There is not, it's been discussed as something that would fit in an
operator alongside (i think it's unlikely it would go into stateful set).
Thanks for the clarity on parallel and the use case.

On Wed, Mar 7, 2018 at 12:25 PM, Alex Robinson <notifications@github.com>
wrote:

> Thanks for chiming in, @smarterclayton <https://github.com/smarterclayton>!
> I don't think that ordered tear down would do much for us today. It would
> only be useful if there was also some sort of "scale-down" hook that would
> allow us to run a command when the size of statefulset is being decreased.
> We could then use that hook to tell the cockroach node to drain all of its
> data onto other cockroach nodes before it shut down, which would allow for
> safe scaling down of multiple nodes if they were done one-by-one.
>
> As things are today, we don't want to drain all of a node's data every
> time it gets stopped, since it'd be incredibly wasteful if the node is
> coming right back up in a matter of seconds or minutes (e.g. due to a
> rolling restart). That means that as of today when the statefulset is
> actually scaled down, we don't create new copies of the data that was on
> that node until a configurable threshold of time passes, at which point we
> consider the node dead. If multiple nodes are brought down in less than
> that amount of time (5 minutes by default) without the admin having
> manually told cockroach to drain the nodes, then unavailability is likely.
>
> Please correct me if I'm wrong about there not being an existing hook for
> statefulset scale-down, though. That would be very useful for us.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/23477#issuecomment-371215283>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p_IFv4K2LWX3Ikd2UcPSS9AxewnYks5tcBgLgaJpZM4SfCMU>
> .
>

a-robinson commented:
I've updated the docs in the README to reflect the minimum version change and new certificate-signing experience on cluster startup. PTAL @mberhault.

a-robinson commented:
As a proposal of sorts, I've added a commit that keeps around old versions of our config files that work on older versions of Kubernetes. Let me know what you think about this idea. I've updated [our release instructions](https://github.com/cockroachdb/cockroach/wiki/Release-process#orchestrator-configurations) with a single command to update all the docker image tags at once to avoid the pain of having to keep these up-to-date manually.

cc @bdarnell
